keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1377 @@
+kind: LeaseCandidate
+metadata:
+  labels:
+    binary-version: "1.29"
+    compatibility-version: "1.29"
+  name: some-custom-controller-0001A
+  namespace: kube-system
+spec:
+  canLeadLease: kube-system/some-custom-controller
+  leaseDurationSeconds: 10
+  renewTime: "2023-12-05T02:33:08.685777Z"
+```
+
+A component "lease candidate" announces candidacy for leadership by specifying
+`spec.canLeadLease` in its lease candidate lease. If the lease expires, the
+component is considered unavailable for leader election purposes.
+
+### Coordinated Election Controller
+
+A new Coordinated Election Controller will reconcile component leader `Lease`s
+(primary resource) and Lease Candidate Leases (secondary resource, changes trigger
+reconciliation of related leader leases).
+
+Coordinated Election Controller reconciliation loop:
+
+- If no leader lease exists for a components:
+  - Elect leader from candidates by preparing a freshly renewed `Lease` with:
+    - `spec.holderIdentity` set to the identity of the elected leader
+    - `coordination.k8s.io/elected-by: leader-election-controller` (to make
+      lease types easy to disambiguate)
+- If there is a better candidate than current leader:
+  - Sets `endofterm: true` on the leader `Lease`, signaling
+    that the leader should stop renewing the lease and yield leadership
+
+```mermaid
+flowchart TD
+   A[Reconcile] --> |Process Leader Lease| B
+   B{Lease Status?} --> |Better Leader Exists| D
+   B --> |Expired/Missing| E
+   D[End Lease Term]
+   E[Elect Leader]
+```
+
+Example of a lease created by Coordinated Election Controller:
+
+```yaml
+apiVersion: coordination.k8s.io/v1
+kind: Lease
+metadata:
+  annotations:
+    coordination.k8s.io/elected-by: coordinated-election-controller
+  name: some-custom-controller
+  namespace: kube-system
+spec:
+  holderIdentity: controller-a
+  leaseDurationSeconds: 10
+  leaseTransitions: 0
+  renewTime: "2023-12-05T18:58:31.295467Z"
+```
+
+The Coordinated Election Controller will run in the kube-apiserver.
+
+In an HA configuration, the Coordinated Leader Election Controller will have its
+own lease similar to how other leader elected controllers behaves today. It will
+be responsible for renewing its own lease and gracefully shutdown if the lease
+is expired. Only one instance of the coordinated leader election controller will
+be active at a time, and this prevents instances of the coordinated leader
+election controller from interfering with each other. Unlike in KCM, the
+coordinated leader election controller must gracefully shutdown and restart as
+it will be running in the kube-apiserver and calling `os.Exit()` is not an
+option.
+
+### Coordinated Lease Lock
+
+A new `resourceLock` type of `coordinatedleases`, and `CoordinatedLeaseLock`
+implementation of `resourcelock.Interface` will be added to client-go that:
+
+- Creates LeaseCandidate Lease when ready to be Leader
+- Renews LeaseCandidate lease very infrequently
```

Comment:
what does this mean? The examples above have 10s.

Reply:
10s is for the leader Lease, LeaseCandidate objects will have an interval of 300s.